### PR TITLE
Add SessionRetryTotal and SessionRetryWait on Android

### PR DESCRIPTION
**Pull request by timwr**
This change adds support for the SessionRetryTotal and SessionRetryWait advanced options.
To make this work the meterpreter payload has to run on a single thread, otherwise we end up spawning multiple sessions (one every SessionRetryWait period).
Verification:
 - [x] Ensure you can still get a reverse_tcp session
 - [ ] Set a non-default value for SessionRetryTotal and SessionRetryWait and ensure the payload waits the SessionRetryWait second period between retries for up to a total of SessionRetryTotal seconds.

### CODE REVIEW

File: androidpayload/app/src/com/metasploit/stage/Payload.java [hunk 1]
```diff
@@ -19,166 +20,172 @@
...
+    }
+
+    private static void reverseHTTP() throws Exception {
+        int checksum;
+        String URI;
+        HttpURLConnection urlConn;
+        String lurl = URL.substring(4).trim();
+
+        while (true) {
+            URI = randomString(4);
```
Comment (hmoore-r7) [inline androidpayload/app/src/com/metasploit/stage/Payload.java]: URIs should always be set by the metasploit-framework side (even a transport change call will pass down the new URL). The INITJ/CONNECT URIs are now much longer as well (30-128 bytes)

**Comment by bcook-r7** [comment]
I'm having trouble in that the HTTP URL doesn't seem to be getting patched in. Instead, I get backtraces like this with http/https:

```
W/System.err( 3018): java.net.ConnectException: /127.0.0.1:4444 - Connection refused
W/System.err( 3018): 	at org.apache.harmony.luni.net.PlainSocketImpl.connect(PlainSocketImpl.java:207)
W/System.err( 3018): 	at org.apache.harmony.luni.net.PlainSocketImpl.connect(PlainSocketImpl.java:183)
W/System.err( 3018): 	at java.net.Socket.startupSocket(Socket.java:705)
W/System.err( 3018): 	at java.net.Socket.tryAllAddresses(Socket.java:150)
W/System.err( 3018): 	at java.net.Socket.<init>(Socket.java:209)
W/System.err( 3018): 	at java.net.Socket.<init>(Socket.java:176)
W/System.err( 3018): 	at com.metasploit.stage.Payload.reverseTCP(Payload.java:130)
W/System.err( 3018): 	at com.metasploit.stage.Payload.startReverseConn(Payload.java:64)
W/System.err( 3018): 	at com.metasploit.stage.Payload.main(Payload.java:53)
W/System.err( 3018): 	at com.metasploit.stage.Payload$1.run(Payload.java:36)
```

tcp works fine though.

**Comment by timwr** [comment]
@bcook-r7 startReverseConn was removed in this change, I suspect you haven't checked out/built branch.

**Comment by bcook-r7** [comment]
Thanks, I had the wrong branch, but an incomplete build. It's looking better.

**Comment by bcook-r7** [comment]
The cert validation looks good. Here is the logcat output with a bad cert, a good one works as expected:
```
W/System.err(31161): javax.net.ssl.SSLHandshakeException: java.security.cert.CertificateException: Invalid certificate
W/System.err(31161): 	at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:322)
W/System.err(31161): 	at com.android.okhttp.Connection.upgradeToTls(Connection.java:201)
W/System.err(31161): 	at com.android.okhttp.Connection.connect(Connection.java:155)
W/System.err(31161): 	at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:276)
W/System.err(31161): 	at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:211)
W/System.err(31161): 	at com.android.okhttp.internal.http.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:382)
```

**Comment by bcook-r7** [comment]
Setting SessionRetryTotal=5 and SessionRetryWait=5 seems intuitively like it would try to stage 5 times over a total period of 25 seconds, but it only attempts once during staging when the payload is first started. If a session connects, it does not seem to ever give up subsequently.

I'm going to go ahead and merge for now so we can get a test payload with actual android binaries and the cert validation piece in place.

**Comment by bcook-r7** [comment]
To clarify, the implementation seems correct, just the variable name we chose always confuses me at first. I should learn to read the descriptions more often :P

**Pull request merged**